boards: convert images to JPEG and reduce image size

gmarull opened the pull request:
The boards folder uses ~142.8 MB, being the largest in the repository.
This is due mostly to board images, which are in most cases not
optimized for web content. This patch tries to address this problem by
converting all pictures to JPEG (quality 75) and by adjusting its size
up to 750 px (the width of the documentation content). Images that
specified a fixed width in rst files are converted down to that value
instead.

With this patch, folder goes down to ~52.6 MB from 142.8 MB (-~63%).
Note that this patch introduces a new set of binary files to git
history, though (bad).

The process has been automated using a Python script (see commit for details)

This also makes me wonder if docs should live in its own repo as many projects do.

**PLEASE REVIEW RESULTS HERE:**
https://builds.zephyrproject.io/zephyr/pr/49024/docs/

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/42356

galak commented:
We really need a CI check for this @stephanosio

stephanosio commented:
> We really need a CI check for this @stephanosio - the file size of image files.

Please file an issue for this if you think this is necessary.

galak commented:
> Please file an issue for this if you think this is necessary.

which repo are we using for such issues?

carlescufi commented:
> > Please file an issue for this if you think this is necessary.
> 
> which repo are we using for such issues?

Main zephyr repo @galak

galak commented:
> > > Please file an issue for this if you think this is necessary.
> > 
> > 
> > which repo are we using for such issues?
> 
> Main zephyr repo @galak

https://github.com/zephyrproject-rtos/zephyr/issues/49026

gmarull committed:
boards: delete unused images

Delete images not referenced in any .rst file.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

aescolar commented on boards/posix/native_posix/doc/index.rst:
```diff
@@ -351,7 +351,7 @@ side-effects.
 ************
 
-.. figure:: native_layers.svg
+.. figure:: native_layers.jpg
```
You don't want to convert vector graphic files. (they tend to be much smaller, they look better, they are editable and may be the source itself for the drawing)

gmarull replied:
will fix that, it's an issue in the script thanks for spotting this one

evgeniy-paltsev reviewed (dismissed):
As I can see some of the images were significantly reduced in size, so now it's really hard to read some captions on the picture. I.e: https://raw.githubusercontent.com/zephyrproject-rtos/zephyr/eccc3235c8bdcf378dbf8c0a41db578c5222614a/boards/arc/hsdk/doc/mikrobus_header.jpg

gmarull commented:
> As I can see some of the images were significantly reduced in size, so now it's really hard to read some captions on the picture. I.e: [`eccc323`/boards/arc/hsdk/doc/mikrobus_header.jpg (raw)](https://raw.githubusercontent.com/zephyrproject-rtos/zephyr/eccc3235c8bdcf378dbf8c0a41db578c5222614a/boards/arc/hsdk/doc/mikrobus_header.jpg)

The script resizes images to the specified width in rst, and if you check the docs, you'll see it actually looks the same:

![image](https://user-images.githubusercontent.com/25011557/184830056-f83247dc-d5df-4624-a4b4-15ae233db84a.png)
https://docs.zephyrproject.org/latest/boards/arc/hsdk/doc/index.html

So, in this case, it looks like the width in the rst needs to be fixed.

gmarull committed:
boards: arc: increase documentation image widths

Some board pictures contain text that is difficult to read with current
width. Increase it to 650px.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
boards: acrn: set fixed image width

Instead of using a relative width, use a fixed size. Because page width
is ~750px, 80% corresponds to ~600px.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

firscity reviewed (changes requested):
Xen Project picture in xenvm docs probably was converted to JPG incorrectly and now is distorted. Please check this, thanks.
https://builds.zephyrproject.io/zephyr/pr/49024/docs/boards/arm64/xenvm/doc/index.html?highlight=xenvm

gmarull commented:
> Xen Project picture in xenvm docs probably was converted to JPG incorrectly and now is distorted. Please check this, thanks. [builds.zephyrproject.io/zephyr/pr/49024/docs/boards/arm64/xenvm/doc/index.html?highlight=xenvm](https://builds.zephyrproject.io/zephyr/pr/49024/docs/boards/arm64/xenvm/doc/index.html?highlight=xenvm)

that's an interesting artifact. Btw, do we really need logos in board docs? Does this add any value?

firscity commented:
> that's an interesting artifact. Btw, do we really need logos in board docs? Does this add any value?

If repo size should be decreased - my opinion is no, we don't. At least we definitely don't want to have distorted images in docs.

gmarull committed:
boards: qemu: delete all QEMU logos

We had a bunch of QEMU logos embedded in all QEMU boards. Logos do not
add much value to the documentation, so just delete them.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
boards: xenvm: remove logo

Adding logos to the documentation doesn't add much value and increases
repository size (due to images).

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
boards: convert images to JPEG and reduce image size

The boards folder uses ~142.8 MB, being the largest in the repository.
This is due mostly to board images, which are in most cases not
optimized for web content. This patch tries to address this problem by
converting all pictures to JPEG (quality 75) and by adjusting its size
up to 750 px (the width of the documentation content). Images that
specified a fixed width in rst files are converted down to that value
instead.

With this patch, folder goes down to ~53.5 MB from 142.8 MB (-~63%).
Note that this patch introduces a new set of binary files to git
history, though (bad).

The process has been automated using this quickly crafted Python script:

```python
from pathlib import Path
import re
import subprocess

def process(doc, image, image_jpeg, size):
    subprocess.run(
        (
	     f"convert {image}"
	     "-background white -alpha remove -alpha off -quality 75"
	     f"-resize {size}\> {image_jpeg}"
	),
        shell=True,
        check=True,
        cwd=doc.parent,
    )
    if image != image_jpeg:
        (doc.parent / image).unlink()

for doc in Path(".").glob("boards/**/*.rst"):
    with open(doc) as f:
        content = ""
        image = None
        for line in f:
            m = re.match(r"^(\s*)\.\. (image|figure):: (.*)$", line)
            if m:
                if image:
                    process(doc, image, image_jpeg, size)

                image = Path(m.group(3))
                if image.suffix not in (".jpg", ".jpeg", ".png"):
                    content += line
                    image = None
                    continue

                image_jpeg = image.parent / (image.stem + ".jpg")
                size = 750
                content += (
                    f"{m.group(1)}.. {m.group(2)}:: {image_jpeg}\n"
                )
            elif image:
                m = re.match(r"\s*:height:\s*[0-9]+.*$", line)
                if m:
                    continue

                m = re.match(r"\s*:width:\s*([0-9]+).*$", line)
                if m:
                    size = min(int(m.group(1)), size)
                    continue

                content += line
                if line == "\n":
                    process(doc, image, image_jpeg, size)
                    image = None
            else:
                content += line

    with open(doc, "w") as f:
        f.write(content)
```

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
boards: neorv32: remove doc logo

Logos do not add much value in board documentation and increase
repository size, so delete it.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

firscity reviewed (dismissed):
Xenvm part LGTM

enjiamai commented on boards/x86/qemu_x86/doc/index.rst:
```diff
@@ -8,13 +8,6 @@ Overview
-   :align: center
-   :alt: Qemu
-
-   Qemu (Credit: qemu.org)
```
Hi @gmarull , just double check it. The qemu logo photo will be removed, right?

gmarull replied:
yes, I removed all instances of it (other logos as well)

povergoing reviewed (dismissed):
LGTM

evgeniy-paltsev commented:
Hi @gmarull ,

> The script resizes images to the specified width in rst, and if you check the docs, you'll see it actually looks the same:
So, in this case, it looks like the width in the rst needs to be fixed.

Ok, I'll fix it in the rst files. May I ask you to postpone changes for ARC boards (`boards/arc/*`) till than?

Thanks!

gmarull commented:
> Hi @gmarull ,
> 
> > The script resizes images to the specified width in rst, and if you check the docs, you'll see it actually looks the same:
> > So, in this case, it looks like the width in the rst needs to be fixed.
> 
> Ok, I'll fix it in the rst files. May I ask you to postpone changes for ARC boards (`boards/arc/*`) till than?
> 
> Thanks!

I have fixed it, please check latest changeset.

evgeniy-paltsev commented:
> I have fixed it, please check latest changeset.

For mikrobus_header.jpg yes, but I also want to fix it for other ARC boards. I.e: [this](https://raw.githubusercontent.com/zephyrproject-rtos/zephyr/75bdfcc360debdaa40b506ee1285c60d9ae35e15/boards/arc/emsdp/doc/emsdp.jpg) doesn't look really redable.

As previously we could simply open original image size by one mouse click in the documentation that wasn't the issue. But as we reducing original image quality as well - now it is an issue.

I'll prepare PR for ARC boards documentation in a few days.

evgeniy-paltsev commented:
BTW, looking at the original issue we trying to solve I'm wondering if we are doing the right thing?

After this PR get merged the git repo will contain both images - the original sized ones and the resized, which came from this PR, so it will **increase** overall repo size.

And (from my understanding) it's the thing we won't be able fix after this PR get merged (if we are't talking about repo history rewrite)

@gmarull what do you think?

gmarull commented:
> BTW, looking at the original issue we trying to solve I'm wondering if we are doing the right thing?
> 
> After this PR get merged the git repo will contain both images - the original sized ones and the resized, which came from this PR, so it will **increase** overall repo size.
> 

Yes, that is actually mentioned in the commit message. Still, there are some advantages like:

- Reduces the repository archive size
- Reduces the size of shallow clones (~130MB down to ~53MB)
- Reduces documentation artifacts size

It is unfortunate that we ended up in such situation because there's no way back as I guess we won't do history re-writes.

> And (from my understanding) it's the thing we won't be able fix after this PR get merged (if we aren't talking about repo history rewrite)
> 
> @gmarull what do you think?

soburi commented:
LGTM

I've checked 
boards/riscv/longan_nano
boards/arm/arduino_mkrzero

soburi reviewed (dismissed):
LGTM

I've checked
boards/riscv/longan_nano
boards/arm/arduino_mkrzero

soburi reviewed (dismissed):
LGTM

erwango reviewed (dismissed):
Didn't parse _all_ STM32 boards, but no issue in the ones I've picked up

gmarull committed:
boards: fvp_baser_aemv8r: remove scale and center images

Now that images are scaled up to the ~width of the page, there's no need
to scale. Also center images to make things prettier.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

cfriedt commented:
Oh shoot - I just saw this. 

Actually, some PNG images are smaller than their JPG counterparts. Not a big deal, but I've considered that in the past prior to converting to JPG.
